File: backend/plan_agent/src/open-canvas/nodes/generateFollowup.ts
import { LangGraphRunnableConfig } from "@langchain/langgraph";
import { getModelFromConfig } from "../../utils.js";
import {
  getArtifactContent,
  isArtifactMarkdownContent,
} from "@opencanvas/shared/utils/artifacts";
import { Reflections } from "@opencanvas/shared/types";
import { ensureStoreInConfig, formatReflections } from "../../utils.js";
import { FOLLOWUP_ARTIFACT_PROMPT } from "../prompts.js";
import {
  OpenCanvasGraphAnnotation,
  OpenCanvasGraphReturnType,
} from "../state.js";
import { ChatOpenAI } from "@langchain/openai";

/**
 * Generate a followup message after generating or updating an artifact.
 */
export const generateFollowup = async (
  state: typeof OpenCanvasGraphAnnotation.State,
  config: LangGraphRunnableConfig
): Promise<OpenCanvasGraphReturnType> => {
  console.log("call generateFollowup()");

  // const smallModel = await getModelFromConfig(config, {
  //   maxTokens: 250,
  //   // We say tool calling is true here because that'll cause it to use a small model
  //   isToolCalling: true,
  // });
  const smallModel = new ChatOpenAI({
    modelName: "Qwen/Qwen2.5-1.5B-Instruct",
    maxTokens: 250,
    temperature: 0.5,
    openAIApiKey: "EMPTY",
    configuration: {
      baseURL: "http://192.168.75.101:8000/v1",
    },
  });

  const store = ensureStoreInConfig(config);
  const assistantId = config.configurable?.assistant_id;
  if (!assistantId) {
    throw new Error("`assistant_id` not found in configurable");
  }
  const memoryNamespace = ["memories", assistantId];
  const memoryKey = "reflection";
  const memories = await store.get(memoryNamespace, memoryKey);
  const memoriesAsString = memories?.value
    ? formatReflections(memories.value as Reflections, {
        onlyContent: true,
      })
    : "No reflections found.";

  const currentArtifactContent = state.artifact
    ? getArtifactContent(state.artifact)
    : undefined;

  const artifactContent = currentArtifactContent
    ? isArtifactMarkdownContent(currentArtifactContent)
      ? currentArtifactContent.fullMarkdown
      : currentArtifactContent.code
    : undefined;

  const formattedPrompt = FOLLOWUP_ARTIFACT_PROMPT.replace(
    "{artifactContent}",
    artifactContent || "No artifacts generated yet."
  )
    .replace("{reflections}", memoriesAsString)
    .replace(
      "{conversation}",
      state._messages
        .map((msg) => `<${msg.getType()}>\n${msg.content}\n</${msg.getType()}>`)
        .join("\n\n")
    );

  // TODO: Include the chat history as well.
  const response = await smallModel.invoke([
    { role: "user", content: formattedPrompt },
  ]);
  console.log("response: ", response);

  return {
    messages: [response],
    _messages: [response],
  };
};

File: backend/ui_agent/src/ui_graph/generate_followup.py
"""Generate a followup message after generating or updating an artifact."""
from typing import Any, Dict

from langchain.schema import LangChainConfig

from opencanvas.shared.utils.artifacts import (
    get_artifact_content,
    is_artifact_markdown_content,
)

from ...utils import (
    ensure_store_in_config,
    format_reflections,
    get_model_from_config,
)
from ..prompts import FOLLOWUP_ARTIFACT_PROMPT


async def generate_followup(
    state: Dict[str, Any],
    config: LangChainConfig,
) -> Dict[str, Any]:
    """Generate a followup message after generating or updating an artifact."""
    print("call generate_followup()")

    small_model = await get_model_from_config(
        config,
        max_tokens=250,
        # We say tool calling is true here because that'll cause it to use a small model
        is_tool_calling=True,
    )

    store = ensure_store_in_config(config)
    assistant_id = config.configurable.get("assistant_id")
    if not assistant_id:
        raise ValueError("`assistant_id` not found in configurable")

    memory_namespace = ["memories", assistant_id]
    memory_key = "reflection"
    memories = await store.get(memory_namespace, memory_key)
    memories_as_string = (
        format_reflections(memories.value, only_content=True)
        if memories and memories.value
        else "No reflections found."
    )

    current_artifact_content = (
        get_artifact_content(state["artifact"]) if state.get("artifact") else None
    )

    artifact_content = (
        current_artifact_content.fullMarkdown
        if current_artifact_content and is_artifact_markdown_content(current_artifact_content)
        else current_artifact_content.code
        if current_artifact_content
        else None
    )

    formatted_prompt = (
        FOLLOWUP_ARTIFACT_PROMPT.replace(
            "{artifactContent}",
            artifact_content or "No artifacts generated yet."
        )
        .replace("{reflections}", memories_as_string)
        .replace(
            "{conversation}",
            "\n\n".join(
                f"<{msg.get_type()}>\n{msg.content}\n</{msg.get_type()}>"
                for msg in state["_messages"]
            )
        )
    )

    # TODO: Include the chat history as well.
    response = await small_model.invoke([
        {"role": "user", "content": formatted_prompt},
    ])
    print("response: ", response)

    return {
        "messages": [response],
        "_messages": [response],
    } 